cpp/src/parquet/page_index.cc

```diff
@@ -184,8 +185,219 @@ class OffsetIndexImpl : public OffsetIndex {
+        offset_index_location->offset - index_read_range_.offset_index->offset;
+    uint32_t length = static_cast<uint32_t>(offset_index_location->length);
+    DCHECK_GE(buffer_offset, 0);
+    DCHECK_LE(buffer_offset + length, index_read_range_.offset_index->length);
+
+    std::shared_ptr<OffsetIndex> offset_index;
+    try {
+      offset_index =
+          OffsetIndex::Make(buffer->data() + buffer_offset, length, properties_);
+    } catch (...) {
+      throw ParquetException("Cannot deserialize offset index for column {}", i);
+    }
+    return offset_index;
+  }
+
+ private:
+  /// The input stream that can perform random access read.
+  ::arrow::io::RandomAccessFile* input_;
+
+  /// The row group metadata to get column chunk metadata.
+  std::shared_ptr<RowGroupMetaData> row_group_metadata_;
+
+  /// Reader properties used to deserialize thrift object.
+  const ReaderProperties& properties_;
+
+  /// File-level decryptor.
+  std::shared_ptr<InternalFileDecryptor> file_decryptor_;
+
+  /// File offsets and sizes of the page Index of all column chunks in the row group.
+  RowGroupIndexReadRange index_read_range_;
+
+  /// Buffer to hold the raw bytes of the page index.
+  /// Will be set lazily when the corresponding page index is accessed for the 1st time.
+  std::shared_ptr<::arrow::Buffer> column_index_buffer_;
+  std::shared_ptr<::arrow::Buffer> offset_index_buffer_;
+};
+
+class PageIndexReaderImpl : public PageIndexReader {
+ public:
+  PageIndexReaderImpl(::arrow::io::RandomAccessFile* input,
+                      std::shared_ptr<FileMetaData> file_metadata,
+                      const ReaderProperties& properties,
+                      std::shared_ptr<InternalFileDecryptor> file_decryptor)
+      : input_(input),
+        file_metadata_(std::move(file_metadata)),
+        properties_(properties),
+        file_decryptor_(std::move(file_decryptor)) {}
+
+  std::shared_ptr<RowGroupPageIndexReader> RowGroup(int i) override {
+    if (i < 0 || i >= file_metadata_->num_row_groups()) {
+      throw ParquetException("Invalid row group ordinal {}", i);
+    }
+    return std::make_shared<RowGroupPageIndexReaderImpl>(
+        input_, file_metadata_->RowGroup(i), properties_, i, file_decryptor_);
+  }
+
+  void WillNeed(const std::vector<int32_t>& row_group_indices,
+                IndexSelection index_selection) override {
+    std::vector<::arrow::io::ReadRange> read_ranges;
+    for (int32_t row_group_ordinal : row_group_indices) {
+      auto read_range = PageIndexReader::DeterminePageIndexRangesInRowGroup(
+          *file_metadata_->RowGroup(row_group_ordinal));
+      if (index_selection.column_index && read_range.column_index.has_value()) {
+        read_ranges.push_back(*read_range.column_index);
+      }
+      if (index_selection.offset_index && read_range.offset_index.has_value()) {
+        read_ranges.push_back(*read_range.offset_index);
+      }
+    }
+    PARQUET_IGNORE_NOT_OK(input_->WillNeed(read_ranges));
```

Comment:
could you add a comment on why this is being ignored.

Reply:
I was thinking that the WillNeed call to the input stream is not required (served as a hint) and then we probably don't care about its failure. But this may hide and postpone the IO issue. Changed it to use `PARQUET_THROW_NOT_OK`.